### PR TITLE
INT-3967: Add contentDescriptor to Object Model

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/INT-3967

### CODE REVIEW

File: spring-integration-core/src/main/java/org/springframework/integration/support/management/graph/IntegrationGraphServer.java [hunk 1]
```diff
@@ -43,6 +43,8 @@
  */
 public class IntegrationGraphServer implements ApplicationContextAware, ApplicationListener<ContextRefreshedEvent> {
 
+	private final float GRAPH_VERSION = 1.0f;
```
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/support/management/graph/IntegrationGraphServer.java]: `static` ?

File: spring-integration-core/src/main/java/org/springframework/integration/support/management/graph/IntegrationGraphServer.java [hunk 2]
```diff
@@ -79,6 +81,17 @@ public void onApplicationEvent(ContextRefreshedEvent event) {
...
 	private synchronized Graph buildGraph() {
+		String implementationVersion = IntegrationGraphServer.class.getPackage().getImplementationVersion();
+		if (implementationVersion == null) {
+			implementationVersion = "unknown - is Spring Integration running from the distribution jar?";
+		}
+		Map<String, Object> descriptor = new HashMap<String, Object>();
+		descriptor.put("provider", "spring-integration");
+		descriptor.put("providerVersion", implementationVersion);
+		descriptor.put("providerFormatVersion", GRAPH_VERSION);
+		if (this.applicationContext.containsBean("applicationNameAndVersion")) {
```
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/support/management/graph/IntegrationGraphServer.java]: That's something weird.
Let allow to inject the `appName` and its `version` into the `IntegrationGraphServer` bean!

Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/support/management/graph/IntegrationGraphServer.java]: I am not sure what you mean by weird.

This way they can use your default graph builder - the bean can be of any type - might be a map, for example.

Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/support/management/graph/IntegrationGraphServer.java]: Yeah... I worried here about Boot's:
```
${spring.application.name}
${application.version}
```
That's why I don't see reason in the separate extra bean, when it really would be better to free ourselves via external injections.
And yes: get a gain, finally, from the Boot environment.

Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/support/management/graph/IntegrationGraphServer.java]: Right - I meant to add code to fall back to that - I think we need to handle non-boot environments.

Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/support/management/graph/IntegrationGraphServer.java]: Both cases can simply be managed by two setters on the `IntegrationGraphServer`.
What is your point?

**Comment by garyrussell** [comment]
Pushed

**Comment by artembilan** [comment]
See the polishing from my side: https://github.com/artembilan/spring-integration/commit/11d60f8803e44391e3dbb006f63897a2a8f4aae5.
The JSON looks like from that test-case:
```
{
  "contentDescriptor": {
    "providerVersion": "unknown - is Spring Integration running from the distribution jar?",
    "providerFormatVersion": 1.0,
    "provider": "spring-integration",
    "name": "testApplication"
  },
  "nodes": [
    {
      "nodeId": 1,
      "name": "nullChannel",
      "stats": null,
      "componentType": "channel",
      "componentName": "nullChannel"
    },
    {
      "nodeId": 2,
      "name": "errorChannel",
      "stats": null,
      "componentType": "publish-subscribe-channel",
      "componentName": "errorChannel"
    },
    {
      "nodeId": 3,
      "name": "_org.springframework.integration.errorLogger",
      "stats": null,
      "componentType": "logging-channel-adapter",
      "componentName": "_org.springframework.integration.errorLogger",
      "output": null,
      "input": "errorChannel"
    }
  ],
  "links": [
    {
      "from": 2,
      "to": 3
    }
  ]
}
```
Pay attention, please, to the duplication around `name` and `componentName`. 
How about remove one of them?

**Comment by garyrussell** [comment]
`name` is the `beanName`; although we always use the beanName in the component name in the framework, we can't guarantee it is all cases since the setter is public; we could, therefore, drop the `componentName`, I guess.

**Comment by artembilan** [comment]
I think that `componentName` does not make so much sense there, since we would like to see the picture directly as a reflection of the beans.
But from other hand we may use `componentName` in the JMX model.
So, for consistency we should remove `name` and represent the logical, user-provided, `componentName` only. Meanwhile the `name` may be just something generated by the `BeanFactory`...

**Comment by garyrussell** [comment]
Pushed (before I read your comment).

**Comment by garyrussell** [comment]
We can't use component name - in the case of channels, the bean name is how we build the links.

File: build.gradle [hunk 3]
```diff
@@ -261,6 +261,12 @@ subprojects { subproject ->
 		archives javadocJar
 	}
 
+	jar {
+		manifest {
+			attributes('Implementation-Version': version)
+		}
```
Comment (artembilan) [inline build.gradle]:
```
Manifest-Version: 1.0
Implementation-Version: 4.3.0.BUILD-SNAPSHOT
```
:+1: 

I think I'm fine with this.
Just curious if other our guys include this to their jars, too.

Comment (artembilan) [inline build.gradle]: Yeah, true.
Even much better!
```
Manifest-Version: 1.0
Created-By: 1.8.0_60 (Oracle Corporation)
Implementation-Title: spring-jms
Implementation-Version: 4.3.0.BUILD-SNAPSHOT
```

Comment (garyrussell) [inline build.gradle]: I looked at the boot code to see how it puts the version in the banner.

I must admit that I didn't look to see whatever else boot puts in the manifest:

We should consider doing the same:

```
Manifest-Version: 1.0
Implementation-Title: Spring Boot
Implementation-Version: 1.3.3.RELEASE
Archiver-Version: Plexus Archiver
Built-By: bamboo
Specification-Vendor: Pivotal Software, Inc.
Specification-Title: Spring Boot
Implementation-Vendor-Id: org.springframework.boot
Created-By: Apache Maven 3.0.5
Build-Jdk: 1.8.0_60
Specification-Version: 1.3.3.RELEASE
Implementation-Vendor: Pivotal Software, Inc.
```

Perhaps not critical for M2, though.

Comment (artembilan) [inline build.gradle]: Yes. Agreed.
Will raise some JIRA on the matter later.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration/commit/2370296b95b5aac740708ad67c3d2e35920aae6e

**Pull request closed without merging**